### PR TITLE
flake on operator state changes during e2e

### PR DESCRIPTION
**Pull request by deads2k**
If any operator changes available, degraded, or progressing on non-disruptive, non-upgrade jobs, then fail.  It's easy to apply elsewhere and easy to make fail, but let's see where we are.

### CODE REVIEW

File: cmd/openshift-tests/e2e.go [hunk 1]
```diff
@@ -821,3 +823,65 @@ func getEventsByPod(events []*monitor.EventInterval) map[string][]*monitor.Event
...
+	}
+
+	// write a passing test to trigger detection of this issue as a flake. Doing this first to try to see how frequent the issue actually is
+	return []*ginkgo.JUnitTestCase{failure, success}
+}
+
+func testOperatorState(interestingCondition string, events []*monitor.EventInterval) []string {
+	failures := []string{}
+
+	clusterOperatorToEvents := getEventsByOperator(events)
```
Comment (wking) [inline cmd/openshift-tests/e2e.go:863]: library-go-based operators will event on ClusterOperator changes, but I'm not sure non-library-go operators will event (consistently) on those or not.

Comment (deads2k) [reply cmd/openshift-tests/e2e.go:863]: > library-go-based operators will event on ClusterOperator changes, but I'm not sure non-library-go operators will event (consistently) on those or not.

this wires to e2e monitor events, not real events.  Also, non-library-go operators could learn.... :)

File: pkg/monitor/operator.go [hunk 2]
```diff
@@ -53,6 +78,12 @@ func startClusterOperatorMonitoring(ctx context.Context, m Recorder, client conf
 					if c.Type == configv1.OperatorDegraded && c.Status == configv1.ConditionTrue {
 						level = Error
 					}
+					if c.Type == configv1.OperatorAvailable && c.Status == configv1.ConditionFalse {
+						level = Error
+					}
+					if c.Type == configv1.OperatorProgressing && c.Status == configv1.ConditionTrue {
```
Comment (smarterclayton) [inline pkg/monitor/operator.go:84]: I would prefer not to make this an error

Comment (deads2k) [reply pkg/monitor/operator.go:84]: > I would prefer not to make this an error

Updated to warning for now.

**Comment by deads2k** [comment]
/hold 

until we confirm we get the right test cases.  serial looks like it ought to trip on this

**Comment by deads2k** [comment]
/hold cancel

The tests are now flaking, with good messages.

File: cmd/openshift-tests/e2e.go (hunk 1 above)
Comment (smarterclayton) [reply cmd/openshift-tests/e2e.go:863]: i do think we could add an OLM monitor addon - olm isn't as sophisticated at CO unfortuntaely.

File: pkg/monitor/operator.go [hunk 3]
```diff
@@ -47,12 +79,18 @@ func startClusterOperatorMonitoring(ctx context.Context, m Recorder, client conf
 					case len(c.Message) > 0:
 						msg = fmt.Sprintf("condition/%s status/%s changed: %s", c.Type, c.Status, c.Message)
 					default:
-						msg = fmt.Sprintf("condition/%s status/%s changed", c.Type, c.Status)
+						msg = fmt.Sprintf("condition/%s status/%s changed: ", c.Type, c.Status)
```
Comment (smarterclayton) [inline pkg/monitor/operator.go:82]: why?

Comment (deads2k) [reply pkg/monitor/operator.go:82]: > why?

laziness in easy machine parsing above

File: cmd/openshift-tests/e2e.go [hunk 4]
```diff
@@ -6,6 +6,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/openshift/origin/pkg/synthetictests"
```
Comment (mfojtik) [inline cmd/openshift-tests/e2e.go:9]: nit: order

File: cmd/openshift-tests/e2e.go [hunk 5]
```diff
@@ -821,3 +825,113 @@ func getEventsByPod(events []*monitor.EventInterval) map[string][]*monitor.Event
...
+
+	knownOperators := allOperators(events)
+	eventsByOperator := getEventsByOperator(events)
+	for _, condition := range []string{"Available", "Degraded", "Progressing"} {
+		for _, operatorName := range knownOperators.List() {
+			bzComponent := synthetictests.GetBugzillaComponentForOperator(operatorName)
+			if bzComponent == "Unknown" {
+				bzComponent = operatorName
+			}
+			testName := fmt.Sprintf("[bz-%v] clusteroperator/%v should not change condition/%v", bzComponent, operatorName, condition)
```
Comment (mfojtik) [inline cmd/openshift-tests/e2e.go:840]: would it help to clarify this further like "... should not change condition after installation is complete" ?

**Pull request merged**